File: 2353. Design a Food Rating System.cpp
class FoodRatings {
    struct Pair {
        int rating;
        string food;
        Pair(int r, string f) : rating(r), food(f) {}
    };

    struct cmp {
        bool operator()(const Pair& a, const Pair& b) const {
            if (a.rating != b.rating)
                return a.rating < b.rating; 
            return a.food > b.food;         
        }
    };

    unordered_map<string, priority_queue<Pair, vector<Pair>, cmp>> hm; 
    unordered_map<string, int> foodRating;                       
    unordered_map<string, string> fToC;                               

public:
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        int n = foods.size();
        for (int i = 0; i < n; i++) {
            hm[cuisines[i]].push(Pair(ratings[i], foods[i]));
            fToC[foods[i]] = cuisines[i];
            foodRating[foods[i]] = ratings[i];
        }
    }

    void changeRating(string food, int newRating) {
        foodRating[food] = newRating;
        hm[fToC[food]].push(Pair(newRating, food)); 
    }

    string highestRated(string cuisine) {
        auto& pq = hm[cuisine];
        while (!pq.empty() && pq.top().rating != foodRating[pq.top().food])
            pq.pop();
        return pq.top().food;
    }
};

File: 2353. Design a Food Rating System.java
class FoodRatings {
    class Pair{
        int rating;
        String food;
        public Pair(int rating, String food){
            this.rating = rating;
            this.food = food;
        }
    }
    HashMap<String, PriorityQueue<Pair>> hm = new HashMap<>();
    HashMap<String, Integer> foodRating = new HashMap<>();
    HashMap<String, String> fToC = new HashMap<>();
    public FoodRatings(String[] foods, String[] cuisines, int[] ratings) {
        int n = foods.length;
        for(int i = 0; i < n; i++){
            if(hm.containsKey(cuisines[i])) hm.get(cuisines[i]).offer(new Pair(ratings[i], foods[i]));
            else{
                PriorityQueue<Pair> temp = new PriorityQueue<>((a, b) -> a.rating != b.rating ? Integer.compare(b.rating, a.rating) : a.food.compareTo(b.food));
                temp.offer(new Pair(ratings[i], foods[i]));
                hm.put(cuisines[i], temp);
            }
            fToC.put(foods[i], cuisines[i]);
            foodRating.put(foods[i], ratings[i]);
        }
    }
    
    public void changeRating(String food, int newRating) {
        foodRating.put(food, newRating);
        hm.get(fToC.get(food)).offer(new Pair(newRating, food));
    }
    
    public String highestRated(String cuisine) {
        PriorityQueue<Pair> pq = hm.get(cuisine);
        while(pq.peek().rating != foodRating.get(pq.peek().food)) pq.poll();

        return pq.peek().food;
    }
}

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings obj = new FoodRatings(foods, cuisines, ratings);
 * obj.changeRating(food,newRating);
 * String param_2 = obj.highestRated(cuisine);
 */